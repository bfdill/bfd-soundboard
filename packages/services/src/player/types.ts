import { ChildProcess } from 'child_process'

export type DefaultPlayer =
  | 'mplayer'
  | 'afplay'
  | 'mpg123'
  | 'mpg321'
  | 'play'
  | 'omxplayer'
  | 'aplay'
  | 'cmdmp3'
export type What = string | string[]
export type PlayerOptions = {
  player: any
  players?: DefaultPlayer | DefaultPlayer[] | string[]
}
export type PlayerArgs = { [playerName: string]: string[] }
export type PlayOptions = PlayerArgs & { stdio: 'ignore' }
export type Next = Function
export interface IPlayer {
  play: (what: What, options?: PlayOptions | Next, next?: Next) => ChildProcess
}
export type createPlayer = (opts?: PlayerOptions) => IPlayer
