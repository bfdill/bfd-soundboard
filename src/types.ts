import * as Koa from 'koa'
import * as KoaRouter from 'koa-router'
import { Server } from 'http'

export interface IApplication {
  listen: (port: number) => Server
}

export type ApplicationState = {
  correlationId: string
}
export interface IApplicationContext
  extends Koa.ParameterizedContext<ApplicationState> {}
export type ApplicationMiddleware = Koa.Middleware<
  ApplicationState,
  IApplicationContext
>
export type ApplicationKoa = Koa<ApplicationState, IApplicationContext>
export type ApplicationRouter = KoaRouter<ApplicationState, IApplicationContext>
