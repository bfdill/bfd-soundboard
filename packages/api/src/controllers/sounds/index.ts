import * as KoaRouter from 'koa-router'
import { NOT_FOUND, OK } from 'http-status-codes'
import { ISoundService, SoundService } from '@talkboy-9000/sounds'

import { IApplicationContext, ApplicationState } from '../../types'

export interface ISoundsController {
  get: (ctx: IApplicationContext) => void
}

export class SoundsController implements ISoundsController {
  private static instance: ISoundsController | undefined

  constructor(protected readonly soundService: ISoundService) {}

  get = async (context: IApplicationContext) => {
    const logger = context.logger.child({
      controller: {
        name: 'SoundsController',
        method: 'get'
      }
    })
    const sounds = this.soundService.getSounds(logger)

    context.body = { sounds }

    if (sounds === null) {
      logger.warn('no sounds from service')
      context.status = NOT_FOUND
      return
    }

    logger.info({
      soundsLength: sounds.length
    })
    context.status = OK
  }

  static getInstance(): ISoundsController {
    if (this.instance !== undefined) return this.instance

    this.instance = new SoundsController(SoundService.getInstance())

    return this.instance
  }
}

export const getSoundsRouter = () =>
  new KoaRouter<ApplicationState, IApplicationContext>().get(
    '/',
    SoundsController.getInstance().get
  )
