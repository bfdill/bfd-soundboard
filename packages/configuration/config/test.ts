import { DeepPartial } from '@talkboy-9000/utils'

import { FullConfig } from '../src/configuration.types'

const config: DeepPartial<FullConfig> = {}

export default config
