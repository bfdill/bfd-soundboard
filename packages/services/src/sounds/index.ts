import { join } from 'path'

export const PATH_TO_SOUNDS = join(process.cwd(), 'audio')
