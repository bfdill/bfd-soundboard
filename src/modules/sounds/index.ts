import { join, resolve, sep, basename } from 'path'
import * as winston from 'winston'
import * as sane from 'sane'
import { sync } from 'glob'
import { getAppLogger } from '../../app'

export const PATH_TO_SOUNDS = join(process.cwd(), 'audio')

export type Sound = {
  id: string
  filename: string
  basename: string
}

export interface ISoundService {
  getBySoundId(soundId: string, parentLogger: winston.Logger): Sound | undefined
  getSounds: (parentLogger: winston.Logger) => Sound[]
  isPathValid: (filename: string, parentLogger: winston.Logger) => boolean
}

export class SoundService implements ISoundService {
  private static instance: ISoundService | undefined
  public readonly FILE_GLOB = '**/*.mp3'

  protected sounds: Sound[] = []
  protected readonly watcher: sane.Watcher

  constructor(
    public readonly pathToSounds: string,
    protected readonly saneFunction: typeof sane,
    protected readonly logger: winston.Logger
  ) {
    this.addSounds()
    this.watcher = this.createWatch()
  }

  addSounds = () => {
    this.logger.info('add sounds')

    sync(join(this.pathToSounds, this.FILE_GLOB)).forEach(this.addSound)
  }

  addSound = (filename: string) => {
    this.logger.debug({ filename })

    this.sounds = [
      ...this.sounds,
      {
        filename,
        id: filename,
        basename: basename(filename)
      }
    ]
      .sort((a, b) => a.filename.localeCompare(b.filename))
      .map((v, i) => {
        v.id = `${i}`
        return v
      })

    this.logger.silly({
      filename,
      sounds: this.sounds
    })
  }

  createWatch = (): sane.Watcher => {
    return this.saneFunction(this.pathToSounds, { glob: this.FILE_GLOB })
      .on('add', (path: string) => {
        this.addSound(path)

        this.logger.debug({
          path,
          message: 'watch:add'
        })
      })
      .on('delete', (path: string) => {
        this.removeSound(path)

        this.logger.debug({
          path,
          message: 'watch:delete'
        })
      })
  }

  getBySoundId = (
    soundId: string,
    parentLogger: winston.Logger
  ): Sound | undefined => {
    const result = this.sounds.find(s => s.id === soundId)
    parentLogger
      .child({
        service: {
          name: 'SoundService',
          method: 'getBySoundId'
        }
      })
      .debug({
        result,
        soundId,
        message: `getBySoundId(${soundId})`
      })
    return result
  }

  getSounds = (parentLogger: winston.Logger): Sound[] => {
    parentLogger
      .child({
        service: {
          name: 'SoundService',
          method: 'getSounds'
        }
      })
      .silly({ sounds: this.sounds })
    return this.sounds
  }

  isPathValid = (filename: string, parentLogger: winston.Logger): boolean => {
    const absCandidate = resolve(filename) + sep
    const result =
      absCandidate.substring(0, PATH_TO_SOUNDS.length) === PATH_TO_SOUNDS
    parentLogger
      .child({
        service: {
          name: 'SoundService',
          method: 'isPathValid'
        }
      })
      .debug({
        filename,
        result,
        message: `isPathValid(${filename})`,
        pathToSounds: PATH_TO_SOUNDS
      })
    return result
  }

  removeSound = (filename: string) => {
    this.logger.debug(`removeSound(${filename})`)

    const soundIndex = this.sounds.findIndex(s => s.filename === filename)

    if (soundIndex === -1) return

    this.sounds = this.sounds
      .slice(0, soundIndex)
      .concat(this.sounds.slice(soundIndex + 1))
      .sort((a, b) => `${a.filename}`.localeCompare(b.filename))
      .map((v, i) => {
        v.id = `${i}`
        return v
      })

    this.logger.silly({
      message: `removeSound(${filename})`,
      sounds: this.sounds
    })
  }

  static getInstance(): ISoundService {
    if (this.instance !== undefined) return this.instance

    this.instance = new SoundService(
      PATH_TO_SOUNDS,
      sane,
      getAppLogger().child({
        service: {
          name: 'SoundService'
        }
      })
    )

    return this.instance
  }
}
