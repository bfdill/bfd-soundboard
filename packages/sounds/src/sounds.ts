import * as sane from 'sane'
import * as winston from 'winston'
import { sync } from 'glob'
import { join, resolve, sep, basename } from 'path'
import { ISoundService, LOGGER_META } from './sounds.types'
import { Sound } from '@talkboy-9000/models'
import {
  Sounds as SoundConfig,
  ConfigurationService
} from '@talkboy-9000/configuration'

export class SoundService implements ISoundService {
  private static instance: ISoundService | undefined

  protected sounds: Sound[] = []
  protected readonly watcher: sane.Watcher

  constructor(
    public readonly config: SoundConfig,
    protected readonly saneFunction: typeof sane,
    protected readonly logger: winston.Logger
  ) {
    this.addSounds()
    this.watcher = this.createWatch()
  }

  addSounds = () => {
    this.logger.info('add sounds')

    sync(join(this.config.PathToSounds, this.config.FileGlob)).forEach(
      this.addSound
    )
  }

  addSound = (filename: string) => {
    this.logger.debug({ filename })

    this.sounds = [
      ...this.sounds,
      {
        filename,
        id: filename,
        basename: basename(filename)
      }
    ]
      .sort((a, b) => a.filename.localeCompare(b.filename))
      .map((v, i) => {
        v.id = `${i}`
        return v
      })

    this.logger.silly({
      filename,
      sounds: this.sounds
    })
  }

  private createWatch = (): sane.Watcher => {
    return this.saneFunction(this.config.PathToSounds, {
      glob: this.config.FileGlob
    })
      .on('add', (path: string) => {
        this.addSound(path)

        this.logger.debug({
          path,
          message: 'watch:add'
        })
      })
      .on('delete', (path: string) => {
        this.removeSound(path)

        this.logger.debug({
          path,
          message: 'watch:delete'
        })
      })
  }

  getBySoundId = (
    soundId: string,
    parentLogger: winston.Logger
  ): Sound | undefined => {
    const result = this.sounds.find(s => s.id === soundId)
    parentLogger
      .child({
        service: {
          name: 'SoundService',
          method: 'getBySoundId'
        }
      })
      .debug({
        result,
        soundId,
        message: `getBySoundId(${soundId})`
      })
    return result
  }

  getSounds = (parentLogger: winston.Logger): Sound[] => {
    parentLogger
      .child({
        service: {
          name: 'SoundService',
          method: 'getSounds'
        }
      })
      .silly({ sounds: this.sounds })
    return this.sounds
  }

  isPathValid = (filename: string, parentLogger: winston.Logger): boolean => {
    const absCandidate = resolve(filename) + sep
    const result =
      absCandidate.substring(0, this.config.PathToSounds.length) ===
      this.config.PathToSounds

    parentLogger
      .child({
        service: {
          name: 'SoundService',
          method: 'isPathValid'
        }
      })
      .debug({
        filename,
        result,
        message: `isPathValid(${filename})`,
        pathToSounds: this.config.PathToSounds
      })

    return result
  }

  removeSound = (filename: string) => {
    this.logger.debug(`removeSound(${filename})`)

    const soundIndex = this.sounds.findIndex(s => s.filename === filename)

    if (soundIndex === -1) return

    this.sounds = this.sounds
      .slice(0, soundIndex)
      .concat(this.sounds.slice(soundIndex + 1))
      .sort((a, b) => `${a.filename}`.localeCompare(b.filename))
      .map((v, i) => {
        v.id = `${i}`
        return v
      })

    this.logger.silly({
      message: `removeSound(${filename})`,
      sounds: this.sounds
    })
  }

  static getInstance(): ISoundService {
    if (this.instance !== undefined) return this.instance

    this.instance = new SoundService(
      ConfigurationService.getInstance().getSounds(),
      sane,
      winston.createLogger({ defaultMeta: LOGGER_META })
    )

    return this.instance
  }
}
