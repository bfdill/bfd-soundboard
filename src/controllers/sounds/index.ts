import * as KoaRouter from 'koa-router'
import { NOT_FOUND, OK } from 'http-status-codes'

import { IApplicationContext, ApplicationState } from '../../types'
import { ISoundService, soundService } from '../../modules/sounds'

export interface ISoundsController {
  get: (ctx: IApplicationContext) => void
}

export class SoundsController implements ISoundsController {
  private static instance: ISoundsController | undefined

  constructor(protected readonly soundService: ISoundService) {}

  get = async (context: IApplicationContext) => {
    const logger = context.logger.child({
      controller: {
        name: 'SoundsController',
        method: 'get'
      }
    })
    const sounds = this.soundService.getSounds()

    context.body = { sounds }

    if (sounds === null) {
      logger.warn('no sounds from service')
      context.status = NOT_FOUND
      return
    }

    logger.info({
      soundsLength: sounds.length
    })
    context.status = OK
  }

  static getInstance(): ISoundsController {
    if (this.instance !== undefined) return this.instance

    this.instance = new SoundsController(soundService)

    return this.instance
  }
}

export const getSoundsRouter = () =>
  new KoaRouter<ApplicationState, IApplicationContext>().get(
    '/',
    SoundsController.getInstance().get
  )
