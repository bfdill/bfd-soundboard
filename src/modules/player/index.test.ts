import { createPlayer, IPlayer } from './types';
import { PlayerService, IPlayerService } from '.';
import { ISoundService } from '../sounds';
import winston = require('winston');
import { WinstonJestTransport } from '../winston-jest/index.test';

// tsd creation fail
const createPlayer: createPlayer = require('play-sound')

describe('modules -> player', () => {
  const mockIsPathValid: jest.Mock = jest.fn()
  const transport: WinstonJestTransport = new WinstonJestTransport()
  const player: IPlayer = {
    play: jest.fn()
  }
  const soundService: ISoundService = {
    getBySoundId: jest.fn(),
    getSounds: jest.fn(),
    isPathValid: mockIsPathValid
  }
  const logger = winston.createLogger({
    transports: [transport]
  })
  const validateLogging = () => {
    transport.mock.mock.calls.forEach(call => {
      expect(call).toMatchSnapshot()
    })
  }
  let playerService: IPlayerService


  beforeEach(() => {
    mockIsPathValid.mockReturnValue(true)
    playerService = new PlayerService(player, soundService, logger)
  })

  test('rejects on invalid file', async () => {
    expect.assertions(3)
    mockIsPathValid.mockReturnValue(false)

    await expect(playerService.playFile('this-file-totes-exists.mp3'))
      .rejects
      .toMatchSnapshot()

    validateLogging()
  })
})
