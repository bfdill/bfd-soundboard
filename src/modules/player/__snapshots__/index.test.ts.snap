// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`modules -> player getPlayerServiceLogger 1`] = `
Array [
  Object {
    "level": "info",
    "message": "✅test works✅",
    "service": "PlayerService",
    Symbol(level): "info",
    Symbol(message): "{\\"message\\":\\"✅test works✅\\",\\"level\\":\\"info\\",\\"service\\":\\"PlayerService\\"}",
  },
]
`;

exports[`modules -> player has known exports 1`] = `
Array [
  "getPlayerServiceLogger",
  "PlayerService",
]
`;

exports[`modules -> player playerService playback happy path 1`] = `undefined`;

exports[`modules -> player playerService playback happy path 2`] = `
Array [
  Object {
    "level": "info",
    "message": Object {
      "filename": "this-file-totes-exists.mp3",
      "method": "playFile",
      "state": Object {
        "correlationId": "1234567890",
      },
    },
    Symbol(level): "info",
    Symbol(message): "{\\"message\\":{\\"filename\\":\\"this-file-totes-exists.mp3\\",\\"method\\":\\"playFile\\",\\"state\\":{\\"correlationId\\":\\"1234567890\\"}},\\"level\\":\\"info\\"}",
  },
]
`;

exports[`modules -> player playerService playback happy path 3`] = `
Array [
  Object {
    "filename": "this-file-totes-exists.mp3",
    "level": "debug",
    "message": "playback success",
    "method": "playFile",
    "state": Object {
      "correlationId": "1234567890",
    },
    Symbol(level): "debug",
    Symbol(message): "{\\"filename\\":\\"this-file-totes-exists.mp3\\",\\"method\\":\\"playFile\\",\\"state\\":{\\"correlationId\\":\\"1234567890\\"},\\"message\\":\\"playback success\\",\\"level\\":\\"debug\\"}",
  },
]
`;

exports[`modules -> player playerService rejects on invalid file 1`] = `"filename path is invalid"`;

exports[`modules -> player playerService rejects on invalid file 2`] = `
Array [
  Object {
    "level": "info",
    "message": Object {
      "filename": "this-file-totes-exists.mp3",
      "method": "playFile",
      "state": Object {
        "correlationId": "1234567890",
      },
    },
    Symbol(level): "info",
    Symbol(message): "{\\"message\\":{\\"filename\\":\\"this-file-totes-exists.mp3\\",\\"method\\":\\"playFile\\",\\"state\\":{\\"correlationId\\":\\"1234567890\\"}},\\"level\\":\\"info\\"}",
  },
]
`;

exports[`modules -> player playerService rejects on invalid file 3`] = `
Array [
  Object {
    "filename": "this-file-totes-exists.mp3",
    "level": "error",
    "message": "filename path is invalid",
    "method": "playFile",
    "state": Object {
      "correlationId": "1234567890",
    },
    Symbol(level): "error",
    Symbol(message): "{\\"filename\\":\\"this-file-totes-exists.mp3\\",\\"message\\":\\"filename path is invalid\\",\\"method\\":\\"playFile\\",\\"state\\":{\\"correlationId\\":\\"1234567890\\"},\\"level\\":\\"error\\"}",
  },
]
`;

exports[`modules -> player playerService rejects on playback error 1`] = `
Object {
  "error": "KHAAAAAAN!",
}
`;

exports[`modules -> player playerService rejects on playback error 2`] = `
Array [
  Object {
    "level": "info",
    "message": Object {
      "filename": "this-file-totes-exists.mp3",
      "method": "playFile",
      "state": Object {
        "correlationId": "1234567890",
      },
    },
    Symbol(level): "info",
    Symbol(message): "{\\"message\\":{\\"filename\\":\\"this-file-totes-exists.mp3\\",\\"method\\":\\"playFile\\",\\"state\\":{\\"correlationId\\":\\"1234567890\\"}},\\"level\\":\\"info\\"}",
  },
]
`;

exports[`modules -> player playerService rejects on playback error 3`] = `
Array [
  Object {
    "error": Object {
      "error": "KHAAAAAAN!",
    },
    "filename": "this-file-totes-exists.mp3",
    "level": "debug",
    "message": "playback failure",
    "method": "playFile",
    "state": Object {
      "correlationId": "1234567890",
    },
    Symbol(level): "debug",
    Symbol(message): "{\\"error\\":{\\"error\\":\\"KHAAAAAAN!\\"},\\"filename\\":\\"this-file-totes-exists.mp3\\",\\"method\\":\\"playFile\\",\\"state\\":{\\"correlationId\\":\\"1234567890\\"},\\"message\\":\\"playback failure\\",\\"level\\":\\"debug\\"}",
  },
]
`;
